GoogleForCreators/web-stories-wp
# AutoSaving: offline local autosaving (sessionStorage)

## Context



## Summary
This PR implements local autosaving using sessionStorage. Every X seconds, a local autosave is created. When the story is loaded, there's a check for local autosave, and if it exists, the user is prompted to restore it (or dismiss it).


## Relevant Technical Choices



## To-do

- [x] Check how to configure local autosave interval



## User-facing changes
![Screenshot 2022-08-18 at 11 22 21](https://user-images.githubusercontent.com/3294597/185347020-8b2794ca-813a-49b9-bbd6-4fd6fd195890.png)



## Testing Instructions






This PR can be tested by following these steps:

1. Make sure autosaving locally is configured before testing (Default interval is 15 seconds, no need to configure separately)
2. Add a new story, make some changes and save
3. Now make a new change and wait for 15 seconds (or whatever the autosaving interval is configure to be)
4. Reload the story without saving
5. Now a dialogue with prompting to restore the local backup should pop up.
6. Verify that clicking "Restore backup" actually restores the previous unsaved change from step 3. 
7. Verify that clicking "Dismiss" does nothing.
8. Verify that reloading the story after Restoring or Dismissing doesn't show the dialogue anymore (the local autosave should have been deleted after either of the action.
9. Now make changes again and wait for 15 seconds (or whatever the autosaving interval is configured to be)
10. Save the story and reload
11. Verify that no dialogue is displayed (since the saved state is the same as the local baclup.


## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?
Adds tracking for error of saving / getting / deleting sessionStorage data.



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #3842
